pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -191,56 +258,17 @@ func (configgen *ConfigGeneratorImpl) buildSidecarListeners(env *model.Environme
 	noneMode := node.GetInterceptionMode() == model.InterceptionNone
 	listeners := make([]*xdsapi.Listener, 0)
 
-	actualWildcard, actualLocalHostAddress := getActualWildcardAndLocalHost(node)
+	_, actualLocalHostAddress := getActualWildcardAndLocalHost(node)
 
 	if mesh.ProxyListenPort > 0 {
 		inbound := configgen.buildSidecarInboundListeners(env, node, push, proxyInstances)
 		outbound := configgen.buildSidecarOutboundListeners(env, node, push, proxyInstances)
-
-		listeners = append(listeners, inbound...)
-		listeners = append(listeners, outbound...)
-
-		listeners = configgen.generateManagementListeners(node, noneMode, env, listeners)
-
-		tcpProxy := &tcp_proxy.TcpProxy{
-			StatPrefix:       util.BlackHoleCluster,
-			ClusterSpecifier: &tcp_proxy.TcpProxy_Cluster{Cluster: util.BlackHoleCluster},
-		}
-
-		if mesh.OutboundTrafficPolicy.Mode == meshconfig.MeshConfig_OutboundTrafficPolicy_ALLOW_ANY {
-			// We need a passthrough filter to fill in the filter stack for orig_dst listener
-			tcpProxy = &tcp_proxy.TcpProxy{
-				StatPrefix:       util.PassthroughCluster,
-				ClusterSpecifier: &tcp_proxy.TcpProxy_Cluster{Cluster: util.PassthroughCluster},
-			}
-		}
-		var transparent *google_protobuf.BoolValue
-		if node.GetInterceptionMode() == model.InterceptionTproxy {
-			transparent = proto.BoolTrue
-		}
-
-		filter := listener.Filter{
-			Name: xdsutil.TCPProxy,
-		}
-
-		if util.IsXDSMarshalingToAnyEnabled(node) {
-			filter.ConfigType = &listener.Filter_TypedConfig{TypedConfig: util.MessageToAny(tcpProxy)}
-		} else {
-			filter.ConfigType = &listener.Filter_Config{Config: util.MessageToStruct(tcpProxy)}
-		}
-
-		// add an extra listener that binds to the port that is the recipient of the iptables redirect
-		listeners = append(listeners, &xdsapi.Listener{
-			Name:           VirtualListenerName,
-			Address:        util.BuildAddress(actualWildcard, uint32(mesh.ProxyListenPort)),
-			Transparent:    transparent,
-			UseOriginalDst: proto.BoolTrue,
-			FilterChains: []listener.FilterChain{
-				{
-					Filters: []listener.Filter{filter},
-				},
-			},
-		})
+		listeners = configgen.mergeInOutBoundListeners(
+			env, node, push,
+			inbound, outbound,
+			util.IsInboundOutboundListenerSplit(node),
+			false,
+			false)
```

Comment:
this feels unnecessary.
simply check for mesh.ProxyListenPort > 0 or mesh.ProxyInboundListenerPort > 0.. and build inbound..
append virtual listeners similarly

Reply:
Love your idea! It comes to this because it was picked up from the filterchain migration base.
As this in/out split make more sense following your suggestion. Will fix

Reply:
Rewrite the listener builder.